mqtt.go

```diff
@@ -19,598 +24,613 @@ const (
-		select {
-		case <-time.After(keyValueUpdateRetryInterval):
-			logError("[MQTT] did not receive PUBACK for packet ID %d within %v", p.PacketID, keyValueUpdateRetryInterval)
-
-		case ack := <-mc.puback:
-			if ack.PacketID == p.PacketID {
-				logInfo("[MQTT] received PUBACK for packet ID %d", ack.PacketID)
-				return nil
-			}
-			// TBD: Something is really wrong if packet ID does not match. What to do?
-		}
-
-		p.Dup = true
+	if err := conn.stream.Flush(); err != nil {
+		logError("[MQTT] failed to flush %s packet: %s", p.Type(), err.Error())
+		return err
 	}
 
-	return errors.New("key-value update dropped")
+	return nil
 }
 
-func (dc *DeviceContext) openMQTTConn(cmdQueue chan<- *DeviceCommand, evtQueue <-chan *DeviceEvent, evtBulkDataQueue <-chan *DeviceBulkData, evtSyncedBulkDataCh <-chan *DeviceSyncedBulkData, kvSyncQueue chan<- *keyValueSync, kvPushQueue <-chan *keyValueSync) (*mqttConn, error) {
-	mc := &mqttConn{
-		dc:   dc,
-		subs: make(map[string]mqttSubscription),
-	}
+func (conn *mqttConn) closeAllChannels() {
 
-	var config *tls.Config
-	var err error
-	if dc.TLSClientAuth {
-		if dc.TLSConfig == nil {
-			logError("Client certificate is not set: %v", err)
-			return nil, err
-		}
-		config = dc.TLSConfig
+	for _, v := range conn.packIDToChan {
+		close(v)
 	}
 
-	addr := fmt.Sprintf("%s:%d", mqttHost, mqttPort)
+	close(conn.outPktCh)
+	close(conn.connRespCh)
+	close(conn.pingRespCh)
+}
 
-	if mqttTLS {
-		if conn, err := tls.DialWithDialer(mqttDialer, "tcp", addr, config); err == nil {
-			mc.conn = conn
-		} else {
-			logError("MQTT TLS dialer failed: %s", err.Error())
-			return nil, err
-		}
-	} else {
-		if conn, err := mqttDialer.Dial("tcp", addr); err == nil {
-			mc.conn = conn
-		} else {
-			logError("MQTT dialer failed: %s", err.Error())
-			return nil, err
-		}
+func (conn *mqttConn) getPacketID() uint16 {
+	conn.mutex.Lock()
+	defer conn.mutex.Unlock()
+	conn.lastPacketID += 1
+	if conn.lastPacketID == 0 {
+		conn.lastPacketID = 1
 	}
+	return conn.lastPacketID
+}
 
-	mc.stream = packet.NewStream(mc.conn, mc.conn)
-
-	if err := mc.connect(); err != nil {
-		mc.conn.Close()
-		return nil, err
+// Returns the appropriate response channel for the channel type. For pubacks,
+// We will either create it or return the one that we stored previous
+func (conn *mqttConn) getResponseChan(p packet.Packet) mqttPacketChan {
+	//logInfo("[MQTT] Fetching response channel for %s\n", p.Type())
+	switch p.Type() {
+	case packet.CONNECT:
+		return conn.connRespCh
+	case packet.DISCONNECT:
+		// Disconnect is a special case. Server will close, so no response, but
+		// it's a packet type we expect, so don't let it fall to default
+		return nil
+	case packet.CONNACK:
+		return conn.connRespCh
+	case packet.SUBSCRIBE:
+		// Subscribes are blocking, but have packet ID's, and, theoretically,
+		// multiples can be sent, so we enable this at this level.
+		subPkt := p.(*packet.SubscribePacket)
+		subPkt.PacketID = conn.getPacketID()
```

Comment:
As you know SUBSCRIBE packet must contain a PacketID.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349761
I think this should be set in Subscribe method for readability.

Reply:
That makes sense.

Reply:
> Actually, after talking about guaranteed delivery, I think creating the packet ID might be better in the session code

I agree the idea. Packet ID should not be in black-box for the publisher.